File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbtrn01c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBTRN01C — Daily Transaction Validation.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbtrn01c.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbtrn01c implements ProgramRunner {

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        Scenario scn = buildScenario(scenarioName);
        if (scn == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        List<String> displayLines = new ArrayList<>();
        int recordsRead = 0;
        int successfulLookups = 0;
        int failedXref = 0;
        int failedAccount = 0;

        for (DalyTranRecord tran : scn.transactions) {
            recordsRead++;
            String line = tran.toString();
            displayLines.add(line);

            // 2000-LOOKUP-XREF
            CardXrefRecord xref = scn.xrefs.get(tran.tranCardNum);
            if (xref == null) {
                String msg = "CARD NUMBER " + tran.tranCardNum
                        + " COULD NOT BE VERIFIED. SKIPPING TRANSACTION ID-"
                        + tran.tranId;
                displayLines.add(msg);
                failedXref++;
            } else {
                displayLines.add("SUCCESSFUL READ OF XREF");
                displayLines.add("CARD NUMBER: " + xref.xrefCardNum);
                displayLines.add("ACCOUNT ID : " + xref.xrefAcctId);
                displayLines.add("CUSTOMER ID: " + xref.xrefCustId);

                // 3000-READ-ACCOUNT
                AccountRecord account = scn.accounts.get(xref.xrefAcctId);
                if (account == null) {
                    String msg = "ACCOUNT " + xref.xrefAcctId + " NOT FOUND";
                    displayLines.add(msg);
                    failedAccount++;
                } else {
                    displayLines.add("SUCCESSFUL READ OF ACCOUNT FILE");
                    successfulLookups++;
                }
            }
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("RECORDS_READ", String.valueOf(recordsRead));
        out.put("SUCCESSFUL_LOOKUPS", String.valueOf(successfulLookups));
        out.put("FAILED_XREF", String.valueOf(failedXref));
        out.put("FAILED_ACCOUNT", String.valueOf(failedAccount));
        for (int i = 0; i < displayLines.size(); i++) {
            out.put("DISPLAY_" + i, displayLines.get(i));
        }
        return out;
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private Scenario buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            default: return null;
        }
    }

    private Scenario scenarioProcessRecords() {
        Scenario s = new Scenario();
        s.xrefs.put("4111000000001111",
                new CardXrefRecord("4111000000001111", 1, 100000001));
        s.accounts.put(100000001,
                new AccountRecord(100000001, "Y"));
        s.transactions.add(makeTran("TRN0000000000001", "4111000000001111",
                "01", 1, 150.25));
        s.transactions.add(makeTran("TRN0000000000002", "9999999999999999",
                "01", 1, 50.00));
        return s;
    }

    private Scenario scenarioEmptyInput() {
        return new Scenario();
    }

    private static DalyTranRecord makeTran(String id, String cardNum,
                                            String typeCd, int catCd, double amt) {
        DalyTranRecord t = new DalyTranRecord();
        t.tranId = id;
        t.tranCardNum = cardNum;
        t.tranTypeCd = typeCd;
        t.tranCatCd = catCd;
        t.tranAmt = amt;
        return t;
    }

    // ── Inner data types ─────────────────────────────────────────────────

    private static class Scenario {
        final Map<String, CardXrefRecord> xrefs = new HashMap<>();
        final Map<Integer, AccountRecord> accounts = new HashMap<>();
        final List<DalyTranRecord> transactions = new ArrayList<>();
    }

    static class DalyTranRecord {
        String tranId = "";
        String tranTypeCd = "";
        int tranCatCd = 0;
        String tranSource = "";
        String tranDesc = "";
        double tranAmt = 0.0;
        int tranMerchantId = 0;
        String tranMerchantName = "";
        String tranMerchantCity = "";
        String tranMerchantZip = "";
        String tranCardNum = "";
        String tranOrigTs = "";
        String tranProcTs = "";

        @Override
        public String toString() {
            return "DalyTranRecord(tran_id='" + tranId
                    + "', tran_type_cd='" + tranTypeCd
                    + "', tran_cat_cd=" + tranCatCd
                    + ", tran_source='" + tranSource
                    + "', tran_desc='" + tranDesc
                    + "', tran_amt=" + tranAmt
                    + ", tran_merchant_id=" + tranMerchantId
                    + ", tran_merchant_name='" + tranMerchantName
                    + "', tran_merchant_city='" + tranMerchantCity
                    + "', tran_merchant_zip='" + tranMerchantZip
                    + "', tran_card_num='" + tranCardNum
                    + "', tran_orig_ts='" + tranOrigTs
                    + "', tran_proc_ts='" + tranProcTs + "')";
        }
    }

    static class CardXrefRecord {
        final String xrefCardNum;
        final int xrefCustId;
        final int xrefAcctId;

        CardXrefRecord(String cardNum, int custId, int acctId) {
            this.xrefCardNum = cardNum;
            this.xrefCustId = custId;
            this.xrefAcctId = acctId;
        }
    }

    static class AccountRecord {
        final int acctId;
        final String acctActiveStatus;

        AccountRecord(int acctId, String activeStatus) {
            this.acctId = acctId;
            this.acctActiveStatus = activeStatus;
        }
    }
}

File: pipeline/reimpl/python/cbtrn01c.py
"""
Reimplementation of CBTRN01C — CardDemo Daily Transaction Validation/Lookup.

Function: Read daily transactions sequentially, look up the card cross-reference,
then look up the corresponding account. Reports any unresolvable card numbers
or missing accounts.

Note: CBTRN01C does NOT post/write transactions — it is a read-and-verify pass.
Actual posting is done by CBTRN02C.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Callable

from .carddemo_data import TranRecord, CardXrefRecord, AccountRecord, CustomerRecord


# ── Daily transaction record (CVTRA06Y) ──────────────────────────────────────

@dataclass
class DalyTranRecord:
    """Daily transaction input record (RECLN 350)."""
    tran_id: str = ""
    tran_type_cd: str = ""
    tran_cat_cd: int = 0
    tran_source: str = ""
    tran_desc: str = ""
    tran_amt: float = 0.0
    tran_merchant_id: int = 0
    tran_merchant_name: str = ""
    tran_merchant_city: str = ""
    tran_merchant_zip: str = ""
    tran_card_num: str = ""
    tran_orig_ts: str = ""
    tran_proc_ts: str = ""


# ── Result ────────────────────────────────────────────────────────────────────

@dataclass
class LookupResult:
    """Result for a single transaction lookup."""
    tran: DalyTranRecord
    xref: CardXrefRecord | None = None
    account: AccountRecord | None = None
    xref_found: bool = False
    account_found: bool = False
    error_msg: str = ""


@dataclass
class ProcessResult:
    records_read: int = 0
    successful_lookups: int = 0
    failed_xref: int = 0
    failed_account: int = 0
    results: list[LookupResult] = field(default_factory=list)
    display_lines: list[str] = field(default_factory=list)


# ── Repository interfaces ──────────────────────────────────────────────────────

class XrefRepository:
    def __init__(self, xrefs: dict[str, CardXrefRecord]):
        """Keyed by card_num."""
        self._by_card = xrefs

    def find_by_card(self, card_num: str) -> CardXrefRecord | None:
        return self._by_card.get(card_num)


class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> AccountRecord | None:
        return self._accounts.get(acct_id)


# ── Core processing ───────────────────────────────────────────────────────────

def process_daily_transactions(
    transactions: list[DalyTranRecord],
    xref_repo: XrefRepository,
    account_repo: AccountRepository,
    logger: Callable[[str], None] = print,
) -> ProcessResult:
    """Validate daily transactions — mirrors CBTRN01C PROCEDURE DIVISION."""
    result = ProcessResult()
    logger("START OF EXECUTION OF PROGRAM CBTRN01C")

    for tran in transactions:
        result.records_read += 1
        line = str(tran)
        logger(line)
        result.display_lines.append(line)

        lookup = LookupResult(tran=tran)

        # 2000-LOOKUP-XREF
        xref = xref_repo.find_by_card(tran.tran_card_num)
        if xref is None:
            msg = (
                f"CARD NUMBER {tran.tran_card_num} COULD NOT BE VERIFIED. "
                f"SKIPPING TRANSACTION ID-{tran.tran_id}"
            )
            logger(msg)
            result.display_lines.append(msg)
            lookup.error_msg = msg
            result.failed_xref += 1
        else:
            lookup.xref = xref
            lookup.xref_found = True
            logger(f"SUCCESSFUL READ OF XREF")
            logger(f"CARD NUMBER: {xref.xref_card_num}")
            logger(f"ACCOUNT ID : {xref.xref_acct_id}")
            logger(f"CUSTOMER ID: {xref.xref_cust_id}")
            result.display_lines.extend([
                "SUCCESSFUL READ OF XREF",
                f"CARD NUMBER: {xref.xref_card_num}",
                f"ACCOUNT ID : {xref.xref_acct_id}",
                f"CUSTOMER ID: {xref.xref_cust_id}",
            ])

            # 3000-READ-ACCOUNT
            account = account_repo.find(xref.xref_acct_id)
            if account is None:
                msg = f"ACCOUNT {xref.xref_acct_id} NOT FOUND"
                logger(msg)
                result.display_lines.append(msg)
                lookup.error_msg = msg
                result.failed_account += 1
            else:
                lookup.account = account
                lookup.account_found = True
                logger("SUCCESSFUL READ OF ACCOUNT FILE")
                result.display_lines.append("SUCCESSFUL READ OF ACCOUNT FILE")
                result.successful_lookups += 1

        result.results.append(lookup)

    logger("END OF EXECUTION OF PROGRAM CBTRN01C")
    return result


# ── run_vector adapter ───────────────────────────────────────────────────────

def _scenario_process_records():
    xrefs = {
        "4111000000001111": CardXrefRecord(
            xref_card_num="4111000000001111", xref_cust_id=1, xref_acct_id=100000001,
        ),
    }
    accounts = {
        100000001: AccountRecord(acct_id=100000001, acct_active_status="Y"),
    }
    trans = [
        DalyTranRecord(tran_id="TRN0000000000001", tran_card_num="4111000000001111",
                        tran_type_cd="01", tran_cat_cd=1, tran_amt=150.25),
        DalyTranRecord(tran_id="TRN0000000000002", tran_card_num="9999999999999999",
                        tran_type_cd="01", tran_cat_cd=1, tran_amt=50.00),
    ]
    return trans, xrefs, accounts


def _scenario_empty_input():
    return [], {}, {}


_CBTRN01C_SCENARIOS = {
    "PROCESS_RECORDS": _scenario_process_records,
    "EMPTY_INPUT": _scenario_empty_input,
}


def run_vector(inputs: dict) -> dict:
    """Adapter for the differential harness runner contract."""
    scenario_name = str(inputs.get("SCENARIO", "PROCESS_RECORDS")).upper()
    if scenario_name not in _CBTRN01C_SCENARIOS:
        return {"error": f"unknown scenario: {scenario_name!r}"}

    trans, xrefs, accounts = _CBTRN01C_SCENARIOS[scenario_name]()
    xref_repo = XrefRepository(xrefs)
    account_repo = AccountRepository(accounts)

    result = process_daily_transactions(trans, xref_repo, account_repo, logger=lambda _: None)

    out: dict[str, str] = {
        "RECORDS_READ": str(result.records_read),
        "SUCCESSFUL_LOOKUPS": str(result.successful_lookups),
        "FAILED_XREF": str(result.failed_xref),
        "FAILED_ACCOUNT": str(result.failed_account),
    }
    for i, line in enumerate(result.display_lines):
        out[f"DISPLAY_{i}"] = line
    return out
